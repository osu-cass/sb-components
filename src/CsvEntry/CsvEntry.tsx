import * as React from "react";
import {
  generateTooltip,
  ItemRevisionModel,
  CsvRowModel,
  NamespaceModel,
  parseCsv,
  Accordion,
  validItemRevisionModel,
  toCsvText,
  toCsvModel
} from "@src/index";

export interface CsvEntryProps {
  namespaces: NamespaceModel[];
  onItemsUpdate: (items: ItemRevisionModel[]) => void;
  onApply: () => void;
  itemRows: ItemRevisionModel[];
}

export interface CsvEntryState {
  csvInputValue?: string;
  csvData: CsvRowModel[];
}

export class CsvEntry extends React.Component<CsvEntryProps, CsvEntryState> {
  constructor(props: CsvEntryProps) {
    super(props);
    const csvData: CsvRowModel[] = [];
    let csvInputValue: string = "";
    if (props.itemRows.length > 1) {
      props.itemRows.forEach((item, index) => {
        csvData.push({ ...item, index });
      });
      csvInputValue = toCsvText(csvData);
    }
    this.state = {
      csvInputValue,
      csvData
    };
  }

  componentWillReceiveProps(props: CsvEntryProps, state: CsvEntryState) {
    if (
      props.itemRows !== this.props.itemRows ||
      this.props.itemRows.length !== props.itemRows.length
    ) {
      const csvData = toCsvModel(props.itemRows);
      const csvInputValue = toCsvText(csvData);
      this.setState({ csvData, csvInputValue });
    }
  }

  renderHelpText(): JSX.Element {
    return (
      <p>
        Enter namespace, bank key, item id, and section for the namespace which
        has a bank key. Or,<br />
        Enter namespace , item id, and section for the namespace which doesn't
        have a bank key.
      </p>
    );
  }

  renderHelpButton() {
    const helpText: JSX.Element = this.renderHelpText();
    const displayText: JSX.Element = (
      <button
        className="item-nav-btn btn btn-default btn-sm"
        role="button"
        aria-label="Open help text"
      >
        <span className="fa fa-info-circle" aria-hidden="true" />
        Help
      </button>
    );

    return (
      <span className="csv-button-left">
        {generateTooltip({
          helpText,
          displayText,
          displayIcon: undefined
        })}
      </span>
    );
  }

  renderApplyButton() {
    return (
      <span className="csv-button-right">
        <button
          className="item-nav-btn btn btn-primary btn-sm csv-apply-button"
          role="button"
          aria-label="Apply text"
          onClick={this.handleCsvApply}
        >
          <span className="fa fa-check-circle" aria-hidden="true" />
          Apply
        </button>
      </span>
    );
  }

  handleCsvChange(event: React.FormEvent<HTMLTextAreaElement>) {
    const rawCsv: string = event.currentTarget.value;

    this.setState({
      csvInputValue: rawCsv
    });
  }

  handleCsvApply = () => {
    const { csvInputValue } = this.state;
    const { namespaces } = this.props;
    const csvData = parseCsv(csvInputValue, namespaces);

    this.setState({
      csvData
    });
    this.props.onItemsUpdate(csvData);
    this.props.onApply();
  };

  render() {
    return (
      <div className="csv-entry-wrapper">
        <div className="csv-button-row">
          {this.renderHelpButton()}
          {this.renderApplyButton()}
        </div>
        <textarea
          className="csv-text-entry"
          onChange={e => this.handleCsvChange(e)}
          value={this.state.csvInputValue}
        />
      </div>
    );
  }
}
