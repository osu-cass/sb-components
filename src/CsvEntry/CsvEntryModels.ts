import {
  findNamespace,
  ItemRevisionModel,
  NamespaceModel
} from "../ItemBank/ItemBankModels";

import { validItemRevisionModel } from "@src/index";

export interface CsvRowModel extends ItemRevisionModel {
  index: number;
}

export function parseCsv(
  csvValue?: string,
  namespaces?: NamespaceModel[]
): CsvRowModel[] {
  let data: CsvRowModel[] = [];

  if (csvValue && namespaces) {
    const lines = csvValue.split("\n");
    data = parseLines(lines, namespaces);
  }

  return data;
}

export function toCsvModel(items: ItemRevisionModel[]): CsvRowModel[] {
  const csvData: CsvRowModel[] = [];
  items.forEach((item, index) => {
    if (item.itemKey) {
      csvData.push({ ...item, index });
    }
  });

  return csvData;
}

export function toCsvText(items: CsvRowModel[]): string {
  let csvString: string = "";
  items.forEach(item => {
    if (item.itemKey) {
      const itemString = item.bankKey
        ? `${item.namespace},${item.bankKey},${item.itemKey},${item.section}\n`
        : `${item.namespace},${item.itemKey},${item.section}\n`;
      csvString = `${csvString}${itemString}`;
    }
  });

  return csvString;
}

function parseLines(lines: string[], namespaces: NamespaceModel[]) {
  const data: CsvRowModel[] = [];
  let index = 0;

  for (const line of lines) {
    const values = line.split(",");

    const row: CsvRowModel = { index };
    // row.namespace = values[0];
    setNamespace(row, values[0], namespaces);

    if (values.length === 4) {
      setCsvRowWithBankKey(row, values);
    } else if (values.length === 3) {
      setCsvRowWithoutBankKey(row, values, namespaces);
    } else {
      continue;
    }

    row.valid = validItemRevisionModel(row);
    data.push(row);
    index = index + 1;
  }

  return data;
}

function setNamespace(
  row: CsvRowModel,
  namespace: string,
  namespaces: NamespaceModel[]
) {
  const matchedNamespace = findNamespace(namespace, namespaces);
  if (matchedNamespace) {
    row.namespace = matchedNamespace.name;
  }
}

function setCsvRowWithBankKey(row: CsvRowModel, values: string[]) {
  row.hasBankKey = true;
  row.bankKey = +values[1];
  row.itemKey = +values[2];
  row.section = values[3].trim();
}

function setCsvRowWithoutBankKey(
  row: CsvRowModel,
  values: string[],
  namespaces: NamespaceModel[]
) {
  row.hasBankKey = false;
  row.bankKey = getBankKeyByNamespace(row.namespace, namespaces);
  row.itemKey = +values[1];
  row.section = values[2].trim();
}

function getBankKeyByNamespace(
  namespace?: string,
  namespaces?: NamespaceModel[]
): number | undefined {
  let bankKey: number | undefined;

  if (namespaces) {
    if (!namespace) return;
    const matchedNamespace = findNamespace(namespace, namespaces);
    if (matchedNamespace) bankKey = matchedNamespace.bankKey;
  }

  return bankKey;
}
