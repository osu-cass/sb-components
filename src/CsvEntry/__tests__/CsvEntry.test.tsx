import "jsdom-global/register";
import * as React from "react";
import { shallow, mount } from "enzyme";
import {
  CsvEntry,
  CsvEntryProps,
  ItemRevisionModel,
  CsvRowModel
} from "@src/index";

import { CsvEntryState } from "../CsvEntry";
const onCsvTextUpdateMock = jest.fn();
const onItemsUpdateMock = jest.fn();

const item: ItemRevisionModel = {
  itemKey: 187,
  bankKey: 3000,
  section: "math",
  revision: "asfe",
  isaap: ""
};

const item2 = { ...item, bankKey: 2332, revision: "asdf" };

const items: ItemRevisionModel[] = [item, item2];

const csvData1: CsvRowModel = {
  ...item,
  index: 0
};

const csvData2: CsvRowModel = {
  ...item2,
  index: 1
};

const csvData: CsvRowModel[] = [csvData1, csvData2];

const csvEntryProps: CsvEntryProps = {
  namespaces: [],
  onItemsUpdate: (items: ItemRevisionModel[]) => {
    const item = "";
  },
  onApply: jest.fn(),
  itemRows: items
};

const csvEntryState: CsvEntryState = {
  csvInputValue: "hello",
  csvData: [csvData1, csvData2]
};

describe("CsvEntry", () => {
  const wrapper = shallow(<CsvEntry {...csvEntryProps} />);

  it("matches snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("calls handle csv apply", () => {
    const wrapperInstance = wrapper.instance() as CsvEntry;
    wrapper.find(".csv-text-entry").simulate("change", {
      currentTarget: { value: "itemreviewviewer,187,3000,SIW" }
    });
    expect(wrapper).toMatchSnapshot();
  });

  it("calls handle csv change", () => {
    wrapper
      .find(".csv-text-entry")
      .simulate("change", { currentTarget: { value: "hello" } });
    expect(wrapper).toMatchSnapshot();
  });
});
