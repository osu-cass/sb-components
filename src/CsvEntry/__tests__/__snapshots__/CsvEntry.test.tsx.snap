// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CsvEntry calls handle csv blur 1`] = `
<div
  className="csv-entry-wrapper"
>
  <div
    className="help-button"
  >
    <ToolTip
      displayText={
        <button
          aria-label="Open help text"
          className="item-nav-btn btn btn-default btn-sm about-item-btn"
          role="button"
        >
          <span
            aria-hidden="true"
            className="fa fa-info-circle"
          />
          Help
        </button>
      }
      helpText={
        <p>
          Enter bank key and item id for the items you would like to add
        </p>
      }
    />
  </div>
  <textarea
    className="csv-text-entry"
    onBlur={[Function]}
    onChange={[Function]}
  />
</div>
`;

exports[`CsvEntry calls handle csv change 1`] = `
<div
  className="csv-entry-wrapper"
>
  <div
    className="help-button"
  >
    <ToolTip
      displayText={
        <button
          aria-label="Open help text"
          className="item-nav-btn btn btn-default btn-sm about-item-btn"
          role="button"
        >
          <span
            aria-hidden="true"
            className="fa fa-info-circle"
          />
          Help
        </button>
      }
      helpText={
        <p>
          Enter bank key and item id for the items you would like to add
        </p>
      }
    />
  </div>
  <textarea
    className="csv-text-entry"
    onBlur={[Function]}
    onChange={[Function]}
    value="hello"
  />
</div>
`;

exports[`CsvEntry matches snapshot 1`] = `
<div
  className="csv-entry-wrapper"
>
  <div
    className="help-button"
  >
    <ToolTip
      displayText={
        <button
          aria-label="Open help text"
          className="item-nav-btn btn btn-default btn-sm about-item-btn"
          role="button"
        >
          <span
            aria-hidden="true"
            className="fa fa-info-circle"
          />
          Help
        </button>
      }
      helpText={
        <p>
          Enter bank key and item id for the items you would like to add
        </p>
      }
    />
  </div>
  <textarea
    className="csv-text-entry"
    onBlur={[Function]}
    onChange={[Function]}
  />
</div>
`;
